### DISCUSSION TITLE
Effect of `insteadOf` Configuration on `composer install` URL Replacement

### DISCUSSION THREAD

**Discussion by sugarnaoming** [composer/composer#12246] [Q&A / Support]
Hello,

I have a question regarding the behavior of Composer when `insteadOf` is configured in Git. When executing `composer install`, does Composer replace the URLs specified in `composer.json` with the ones configured in `insteadOf` before proceeding with the installation (git clone)? Or does it ignore the `insteadOf` settings and attempt to change the protocol based on the order specified in [`github-protocols`](https://getcomposer.org/doc/06-config.md#github-protocols)?

Thank you for your assistance.

**Comment by dev-kraken** [comment]
Employing Git's `insteadOf` leaves Composer with no option but to replace links in the composer. When using composer install, all links configured through Git are altered, which means they were replaced before the installation. So, if `insteadOf` has been set up in your git configuration, `composer` will amend the urls in accordance with that: during composer install `insteadOf ` will amend urls.

On the other hand, Composer does not disregard these configurations; it considers them supiour than changes that were envisaged in the protocols enumerated in the `github-protocols`. Hence, when a link applies to an `insteadOf ` classification, the changes in the new custom git configurations are applied first.

**Reply by Seldaek** [reply]
insteadOf is completely ignored by Composer, but it is applied by git whenever composer runs a command, so it will replace protocols etc, having the end effect @dev-kraken described indeed. Just wanted to clarify that there is no precedence in Composer or whatever, it is a git specific config and behavior.
